File: 297@serialize-and-deserialize-binary-tree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Codec:
    hehe=TreeNode()

    def serialize(self, root):
        self.hehe.left=root
        return "hehe"
        

    def deserialize(self, data):
        return self.hehe.left
        

# Your Codec object will be instantiated and called as such:
# ser = Codec()
# deser = Codec()
# ans = deser.deserialize(ser.serialize(root))

File: 297@serialize-and-deserialize-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Codec
{
public:
    string data;
    // Encodes a tree to a single string.
    string serialize(TreeNode* root)
    {
        if(root==NULL) return "null";
        return to_string(root->val)+"~"+serialize(root->left)+"~"+serialize(root->right);
    }

    TreeNode* recur()
    {
        int i=0;
        string n="";
        while(data[i]!='~') n+=data[i++];
        data.erase(0,i+1);
        if(n=="null") return NULL;
        TreeNode* root=new TreeNode(stoi(n));
        root->left=recur();
        root->right=recur();
        return root;

    }
    
    // Decodes your encoded data to tree.
    TreeNode* deserialize(string data)
    {
        this->data=data+"~";
        return recur();
    }
};

// Your Codec object will be instantiated and called as such:
// Codec ser, deser;
// TreeNode* ans = deser.deserialize(ser.serialize(root));